README.md
# PyROOT Example on CMS' NanoAOD

Simple example on how to access CMS' NanoAOD files and read some information from it.

**Note:** Assuming (1) CMSSW environment is available, (2) the user have GRID certificate (3) registered to CMS' VO.

**More info:**
- https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookStartingGrid
- https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideLcgAccess

![pyroot_example_nanoaod](https://github.com/ftorresd/pyroot_example_nanoaod/raw/main/content/pyroot_example_nanoaod.gif)

## Download or clone (HTTPS) or clone (SSH)

``` 
wget https://github.com/ftorresd/pyroot_example_nanoaod/archive/refs/heads/main.zip
unzip main.zip
cd pyroot_example_nanoaod-main
```
or 

```
git clone https://github.com/ftorresd/pyroot_example_nanoaod.git
```

or 

```
git clone git@github.com:ftorresd/pyroot_example_nanoaod.git
```

## Setup

Only once per session.

```
source setup_env.sh
```

## Run 

```
python analyser.py
```

## Output

Histograms will be saved at `output_file.root`.


CHANGELOG.md
## v0.0
- Minimally functional code.

## v1.0
- Read files from a TChain
- Source files loading (TChain) isolated in anoter file (`source_files.py`).
- Effective x-sec properly weighted by PDG values.
- Counters scaled to event weight.
- MC content description.
- Filter MC events for muon/electrons decays only.
- Filter o Golden JSON.
- PU reweighting. 
- Nicer plot printing.

stac.py
import ROOT
ROOT.gROOT.SetBatch(True) 
# setup some ROOT stuff
ROOT.PyConfig.DisableRootLogon = True
ROOT.PyConfig.IgnoreCommandLineOptions = False
ROOT.gROOT.SetBatch(True)

# Instructions for CMS plot styles:
# https://twiki.cern.ch/twiki/bin/view/CMS/Internal/PubGuidelines
# https://twiki.cern.ch/twiki/bin/view/CMS/Internal/FigGuidelines

# clean old stuff
import os
os.system("rm -rf outputs_files/*.png")
os.system("rm -rf outputs_files/*.pdf")

# Open a ROOT file 
output_file = ROOT.TFile('output_file.root')

# save the histogram
h_muon_pt_1 = output_file.Get("h_muon_pt_1")
h_muon_phi_1 = output_file.Get("h_muon_phi_1")
h_MET_phi_1 = output_file.Get("h_MET_phi_1")
h_MET_pt_1 = output_file.Get("h_MET_pt_1")
h_nmuon_1 = output_file.Get("h_nmuon_1")
h_muon_pt_2 = output_file.Get("h_muon_pt_2")
h_muon_phi_2 = output_file.Get("h_muon_phi_2")
h_MET_phi_2 = output_file.Get("h_MET_phi_2")
h_MET_pt_2 = output_file.Get("h_MET_pt_2")
h_nmuon_2 = output_file.Get("h_nmuon_2")
h_MmM1 = output_file.Get("h_MmM1")
h_MmM2 = output_file.Get("h_MmM2")

# setting the title
h_muon_pt_1.SetTitle("Data")
h_muon_phi_1.SetTitle("Data")
h_MET_phi_1.SetTitle("Data")
h_MET_pt_1.SetTitle("Data")
h_nmuon_1.SetTitle("Data")
h_MmM1.SetTitle("Data")

h_muon_pt_2.SetTitle("Signal MC")
h_muon_phi_2.SetTitle("Signal MC")
h_MET_phi_2.SetTitle("Signal MC")
h_MET_pt_2.SetTitle("Signal MC")
h_nmuon_2.SetTitle("Signal MC")
h_MmM2.SetTitle("Signal MC")

h_muon_pt_1.SetMarkerStyle(ROOT.kFullCircle)
h_muon_pt_2.SetLineWidth(2)
h_muon_pt_2.SetFillColorAlpha(ROOT.kOrange-2, 0.35)

h_muon_phi_1.SetMarkerStyle(ROOT.kFullCircle)
h_muon_phi_2.SetLineWidth(2)
h_muon_phi_2.SetFillColorAlpha(ROOT.kOrange-2, 0.35)

h_MmM1.SetMarkerStyle(ROOT.kFullCircle)
h_MmM2.SetLineWidth(2)
h_MmM2.SetFillColorAlpha(ROOT.kOrange-2, 0.75)

h_MET_pt_1.SetMarkerStyle(ROOT.kFullCircle)
h_MET_pt_2.SetLineWidth(2)
h_MET_pt_2.SetFillColorAlpha(ROOT.kOrange-2, 0.35)

h_MET_phi_1.SetMarkerStyle(ROOT.kFullCircle)
h_MET_phi_2.SetLineWidth(2)
h_MET_phi_2.SetFillColorAlpha(ROOT.kOrange-2, 0.35)

h_nmuon_1.SetMarkerStyle(ROOT.kFullCircle)
h_nmuon_2.SetLineWidth(2)
h_nmuon_2.SetFillColorAlpha(ROOT.kOrange-2, 0.35)


c1 = ROOT.TCanvas("c1","c1",1200,800)


h_hs_M = ROOT.THStack("h_hs_M","; m_{T} (GeV);Events")
h_hs_muon_pt = ROOT.THStack("h_hs_muon_pt","; Lepton p_{T} (GeV);Events")
h_hs_muon_phi = ROOT.THStack("h_hs_muon_phi","; Lepton #phi;Events")
h_hs_MET_pt = ROOT.THStack("h_hs_MET_pt","; Missing E_{T} (GeV);Events")
h_hs_MET_phi = ROOT.THStack("h_hs_MET_phi","; Missing E_{T} #phi;Events")
h_hs_nmuon = ROOT.THStack("h_hs_nmuon","; Number of muons;Events")

h_hs_M.Add(h_MmM1, "E1")
h_hs_M.Add(h_MmM2, "hist")
h_hs_muon_pt.Add(h_muon_pt_1, "E1")
h_hs_muon_pt.Add(h_muon_pt_2, "hist")
h_hs_muon_phi.Add(h_muon_phi_1, "E1")
h_hs_muon_phi.Add(h_muon_phi_2, "hist")
h_hs_MET_pt.Add(h_MET_pt_1, "E1")
h_hs_MET_pt.Add(h_MET_pt_2, "hist")
h_hs_MET_phi.Add(h_MET_phi_1, "E1")
h_hs_MET_phi.Add(h_MET_phi_2, "hist")
h_hs_nmuon.Add(h_nmuon_1, "E1")
h_hs_nmuon.Add(h_nmuon_2, "hist")


#Drawing Histograms
h_hs_M.Draw("nostack")
c1.Modified()
ROOT.gPad.BuildLegend(0.75,0.75,0.95,0.95,"")
c1.SaveAs("outputs_files/h_hs_M.png")
c1.SaveAs("h_hs_M.pdf")

h_hs_muon_pt.Draw("nostack")
c1.Modified()
ROOT.gPad.BuildLegend(0.75,0.75,0.95,0.95,"")
c1.SaveAs("outputs_files/h_hs_muon_pt.png")
c1.SaveAs("outputs_files/h_hs_muon_pt.pdf")

h_hs_muon_phi.Draw("nostack")
c1.Modified()
ROOT.gPad.BuildLegend(0.75,0.75,0.95,0.95,"")
c1.SaveAs("outputs_files/h_hs_muon_phi.png")
c1.SaveAs("outputs_files/h_hs_muon_phi.pdf")

h_hs_MET_pt.Draw("nostack")
c1.Modified()
ROOT.gPad.BuildLegend(0.75,0.75,0.95,0.95,"")
c1.SaveAs("outputs_files/h_hs_MET_pt.png")
c1.SaveAs("outputs_files/h_hs_MET_pt.pdf")

h_hs_MET_phi.Draw("nostack")
c1.Modified()
ROOT.gPad.BuildLegend(0.75,0.75,0.95,0.95,"")
c1.SaveAs("outputs_files/h_hs_MET_phi.png")
c1.SaveAs("outputs_files/h_hs_MET_phi.pdf")

h_hs_nmuon.Draw("nostack")
c1.Modified()
ROOT.gPad.BuildLegend(0.75,0.75,0.95,0.95,"")
c1.SaveAs("outputs_files/h_hs_nmuon.png")
c1.SaveAs("outputs_files/h_hs_nmuon.pdf")

'''c1.SaveAs("h_hs_M.png")
c1.SaveAs("h_hs_M.pdf")
c1.SaveAs("h_hs_muon_pt.png")
c1.SaveAs("h_hs_muon_pt.pdf")
c1.SaveAs("h_hs_muon_phi.png")
c1.SaveAs("h_hs_muon_phi.pdf")
c1.SaveAs("h_hs_MET_pt.png")
c1.SaveAs("h_hs_MET_pt.pdf")
c1.SaveAs("h_hs_MET_phi.png")
c1.SaveAs("h_hs_MET_phi.pdf")
c1.SaveAs("h_hs_nmuon.png")
c1.SaveAs("h_hs_nmuon.pdf")'''


output_file.Close()

'''# open a file that contains stacked histogras
stacked_file = ROOT.TFile('stacked_file.root', 'RECREATE')

h_hs_M.Write()
h_hs_muon_pt.Write()
h_hs_muon_phi.Write()
h_hs_MET_pt.Write()
h_hs_MET_phi.Write()
h_hs_nmuon.Write()

#close file...
stacked_file.Close()'''



CMS_lumi.py
import ROOT as rt

# CMS_lumi
#   Initiated by: Gautier Hamel de Monchenault (Saclay)
#   Translated in Python by: Joshua Hardenbrook (Princeton)
#   Updated by:   Dinko Ferencek (Rutgers)
#

cmsText     = "CMS";
cmsTextFont   = 61  

writeExtraText = True
extraText   = "Preliminary"
extraTextFont = 52 

lumiTextSize     = 0.6
lumiTextOffset   = 0.2

cmsTextSize      = 0.75
cmsTextOffset    = 0.1

relPosX    = 0.045
relPosY    = 0.035
relExtraDY = 1.2

extraOverCmsTextSize  = 0.76

lumi_13TeV = "20.1 fb^{-1}"
lumi_8TeV  = "19.7 fb^{-1}" 
lumi_7TeV  = "5.1 fb^{-1}"
lumi_sqrtS = ""

drawLogo      = False

def CMS_lumi(pad,  iPeriod,  iPosX ):
    outOfFrame    = False
    if(iPosX/10==0 ): outOfFrame = True

    alignY_=3
    alignX_=2
    if( iPosX/10==0 ): alignX_=1
    if( iPosX==0    ): alignY_=1
    if( iPosX/10==1 ): alignX_=1
    if( iPosX/10==2 ): alignX_=2
    if( iPosX/10==3 ): alignX_=3
    align_ = 10*alignX_ + alignY_

    H = pad.GetWh()
    W = pad.GetWw()
    l = pad.GetLeftMargin()
    t = pad.GetTopMargin()
    r = pad.GetRightMargin()
    b = pad.GetBottomMargin()
    e = 0.025

    pad.cd()

    lumiText = ""
    if( iPeriod==1 ):
        lumiText += lumi_7TeV
        lumiText += " (7 TeV)"
    elif ( iPeriod==2 ):
        lumiText += lumi_8TeV
        lumiText += " (8 TeV)"

    elif( iPeriod==3 ):      
        lumiText = lumi_8TeV 
        lumiText += " (8 TeV)"
        lumiText += " + "
        lumiText += lumi_7TeV
        lumiText += " (7 TeV)"
    elif ( iPeriod==4 ):
        lumiText += lumi_13TeV
        lumiText += " (13 TeV)"
    elif ( iPeriod==7 ):
        if( outOfFrame ):lumiText += "#scale[0.85]{"
        lumiText += lumi_13TeV 
        lumiText += " (13 TeV)"
        lumiText += " + "
        lumiText += lumi_8TeV 
        lumiText += " (8 TeV)"
        lumiText += " + "
        lumiText += lumi_7TeV
        lumiText += " (7 TeV)"
        if( outOfFrame): lumiText += "}"
    elif ( iPeriod==12 ):
        lumiText += "8 TeV"
    elif ( iPeriod==0 ):
        lumiText += lumi_sqrtS
            
    print lumiText

    latex = rt.TLatex()
    latex.SetNDC()
    latex.SetTextAngle(0)
    latex.SetTextColor(rt.kBlack)    
    
    extraTextSize = extraOverCmsTextSize*cmsTextSize
    
    latex.SetTextFont(42)
    latex.SetTextAlign(31) 
    latex.SetTextSize(lumiTextSize*t)    

    latex.DrawLatex(1-r,1-t+lumiTextOffset*t,lumiText)

    if( outOfFrame ):
        latex.SetTextFont(cmsTextFont)
        latex.SetTextAlign(11) 
        latex.SetTextSize(cmsTextSize*t)    
        latex.DrawLatex(l,1-t+lumiTextOffset*t,cmsText)
  
    pad.cd()

    posX_ = 0
    if( iPosX%10<=1 ):
        posX_ =   l + relPosX*(1-l-r)
    elif( iPosX%10==2 ):
        posX_ =  l + 0.5*(1-l-r)
    elif( iPosX%10==3 ):
        posX_ =  1-r - relPosX*(1-l-r)

    posY_ = 1-t - relPosY*(1-t-b)

    if( not outOfFrame ):
        if( drawLogo ):
            posX_ =   l + 0.045*(1-l-r)*W/H
            posY_ = 1-t - 0.045*(1-t-b)
            xl_0 = posX_
            yl_0 = posY_ - 0.15
            xl_1 = posX_ + 0.15*H/W
            yl_1 = posY_
            CMS_logo = rt.TASImage("CMS-BW-label.png")
            pad_logo =  rt.TPad("logo","logo", xl_0, yl_0, xl_1, yl_1 )
            pad_logo.Draw()
            pad_logo.cd()
            CMS_logo.Draw("X")
            pad_logo.Modified()
            pad.cd()          
        else:
            latex.SetTextFont(cmsTextFont)
            latex.SetTextSize(cmsTextSize*t)
            latex.SetTextAlign(align_)
            latex.DrawLatex(posX_, posY_, cmsText)
            if( writeExtraText ) :
                latex.SetTextFont(extraTextFont)
                latex.SetTextAlign(align_)
                latex.SetTextSize(extraTextSize*t)
                latex.DrawLatex(posX_, posY_- relExtraDY*cmsTextSize*t, extraText)
    elif( writeExtraText ):
        if( iPosX==0):
            posX_ =   l +  relPosX*(1-l-r)
            posY_ =   1-t+lumiTextOffset*t

        latex.SetTextFont(extraTextFont)
        latex.SetTextSize(extraTextSize*t)
        latex.SetTextAlign(align_)
        latex.DrawLatex(posX_, posY_, extraText)      

    pad.Update()

tdrstyle.py
import ROOT as rt

def tdrGrid( gridOn):
  tdrStyle.SetPadGridX(gridOn)
  tdrStyle.SetPadGridY(gridOn)

#fixOverlay: Redraws the axis
def fixOverlay(): gPad.RedrawAxis()

def setTDRStyle():
  tdrStyle =  rt.TStyle("tdrStyle","Style for P-TDR")

   #for the canvas:
  tdrStyle.SetCanvasBorderMode(0)
  tdrStyle.SetCanvasColor(rt.kWhite)
  tdrStyle.SetCanvasDefH(600) #Height of canvas
  tdrStyle.SetCanvasDefW(600) #Width of canvas
  tdrStyle.SetCanvasDefX(0)   #POsition on screen
  tdrStyle.SetCanvasDefY(0)


  tdrStyle.SetPadBorderMode(0)
  #tdrStyle.SetPadBorderSize(Width_t size = 1)
  tdrStyle.SetPadColor(rt.kWhite)
  tdrStyle.SetPadGridX(False)
  tdrStyle.SetPadGridY(False)
  tdrStyle.SetGridColor(0)
  tdrStyle.SetGridStyle(3)
  tdrStyle.SetGridWidth(1)

#For the frame:
  tdrStyle.SetFrameBorderMode(0)
  tdrStyle.SetFrameBorderSize(1)
  tdrStyle.SetFrameFillColor(0)
  tdrStyle.SetFrameFillStyle(0)
  tdrStyle.SetFrameLineColor(1)
  tdrStyle.SetFrameLineStyle(1)
  tdrStyle.SetFrameLineWidth(1)
  
#For the histo:
  #tdrStyle.SetHistFillColor(1)
  #tdrStyle.SetHistFillStyle(0)
  tdrStyle.SetHistLineColor(1)
  tdrStyle.SetHistLineStyle(0)
  tdrStyle.SetHistLineWidth(1)
  #tdrStyle.SetLegoInnerR(Float_t rad = 0.5)
  #tdrStyle.SetNumberContours(Int_t number = 20)

  tdrStyle.SetEndErrorSize(2)
  #tdrStyle.SetErrorMarker(20)
  #tdrStyle.SetErrorX(0.)
  
  tdrStyle.SetMarkerStyle(20)
  
#For the fit/function:
  tdrStyle.SetOptFit(1)
  tdrStyle.SetFitFormat("5.4g")
  tdrStyle.SetFuncColor(2)
  tdrStyle.SetFuncStyle(1)
  tdrStyle.SetFuncWidth(1)

#For the date:
  tdrStyle.SetOptDate(0)
  # tdrStyle.SetDateX(Float_t x = 0.01)
  # tdrStyle.SetDateY(Float_t y = 0.01)

# For the statistics box:
  tdrStyle.SetOptFile(0)
  tdrStyle.SetOptStat(0) # To display the mean and RMS:   SetOptStat("mr")
  tdrStyle.SetStatColor(rt.kWhite)
  tdrStyle.SetStatFont(42)
  tdrStyle.SetStatFontSize(0.025)
  tdrStyle.SetStatTextColor(1)
  tdrStyle.SetStatFormat("6.4g")
  tdrStyle.SetStatBorderSize(1)
  tdrStyle.SetStatH(0.1)
  tdrStyle.SetStatW(0.15)
  # tdrStyle.SetStatStyle(Style_t style = 1001)
  # tdrStyle.SetStatX(Float_t x = 0)
  # tdrStyle.SetStatY(Float_t y = 0)

# Margins:
  tdrStyle.SetPadTopMargin(0.05)
  tdrStyle.SetPadBottomMargin(0.13)
  tdrStyle.SetPadLeftMargin(0.16)
  tdrStyle.SetPadRightMargin(0.02)

# For the Global title:

  tdrStyle.SetOptTitle(0)
  tdrStyle.SetTitleFont(42)
  tdrStyle.SetTitleColor(1)
  tdrStyle.SetTitleTextColor(1)
  tdrStyle.SetTitleFillColor(10)
  tdrStyle.SetTitleFontSize(0.05)
  # tdrStyle.SetTitleH(0) # Set the height of the title box
  # tdrStyle.SetTitleW(0) # Set the width of the title box
  # tdrStyle.SetTitleX(0) # Set the position of the title box
  # tdrStyle.SetTitleY(0.985) # Set the position of the title box
  # tdrStyle.SetTitleStyle(Style_t style = 1001)
  # tdrStyle.SetTitleBorderSize(2)

# For the axis titles:

  tdrStyle.SetTitleColor(1, "XYZ")
  tdrStyle.SetTitleFont(42, "XYZ")
  tdrStyle.SetTitleSize(0.06, "XYZ")
  # tdrStyle.SetTitleXSize(Float_t size = 0.02) # Another way to set the size?
  # tdrStyle.SetTitleYSize(Float_t size = 0.02)
  tdrStyle.SetTitleXOffset(0.9)
  tdrStyle.SetTitleYOffset(1.25)
  # tdrStyle.SetTitleOffset(1.1, "Y") # Another way to set the Offset

# For the axis labels:

  tdrStyle.SetLabelColor(1, "XYZ")
  tdrStyle.SetLabelFont(42, "XYZ")
  tdrStyle.SetLabelOffset(0.007, "XYZ")
  tdrStyle.SetLabelSize(0.05, "XYZ")

# For the axis:

  tdrStyle.SetAxisColor(1, "XYZ")
  tdrStyle.SetStripDecimals(True)
  tdrStyle.SetTickLength(0.03, "XYZ")
  tdrStyle.SetNdivisions(510, "XYZ")
  tdrStyle.SetPadTickX(1)  # To get tick marks on the opposite side of the frame
  tdrStyle.SetPadTickY(1)

# Change for log plots:
  tdrStyle.SetOptLogx(0)
  tdrStyle.SetOptLogy(0)
  tdrStyle.SetOptLogz(0)

# Postscript options:
  tdrStyle.SetPaperSize(20.,20.)
  # tdrStyle.SetLineScalePS(Float_t scale = 3)
  # tdrStyle.SetLineStyleString(Int_t i, const char* text)
  # tdrStyle.SetHeaderPS(const char* header)
  # tdrStyle.SetTitlePS(const char* pstitle)

  # tdrStyle.SetBarOffset(Float_t baroff = 0.5)
  # tdrStyle.SetBarWidth(Float_t barwidth = 0.5)
  # tdrStyle.SetPaintTextFormat(const char* format = "g")
  # tdrStyle.SetPalette(Int_t ncolors = 0, Int_t* colors = 0)
  # tdrStyle.SetTimeOffset(Double_t toffset)
  # tdrStyle.SetHistMinimumZero(kTRUE)

  tdrStyle.SetHatchesLineWidth(5)
  tdrStyle.SetHatchesSpacing(0.05)

  tdrStyle.cd()

source_files.py
import ROOT

# Create a Chain of filles to analyze

## MC
mc_chain = ROOT.TChain("Events")
print "--> Adding MC files to chain..."
# Dataset: /WJetsToLNu_TuneCP5_13TeV-madgraphMLM-pythia8/RunIISummer20UL18NanoAODv2-106X_upgrade2018_realistic_v15_L1v1-v1/NANOAODSIM
# List all files:
# dasgoclient -query="file dataset=/WJetsToLNu_TuneCP5_13TeV-madgraphMLM-pythia8/RunIISummer20UL18NanoAODv2-106X_upgrade2018_realistic_v15_L1v1-v1/NANOAODSIM"

# mc_chain.Add("root://cms-xrd-global.cern.ch///store/mc/RunIISummer20UL18NanoAODv2/WJetsToLNu_TuneCP5_13TeV-madgraphMLM-pythia8/NANOAODSIM/106X_upgrade2018_realistic_v15_L1v1-v1/00000/B81103CC-8168-3F46-AEC4-6CEF3012FC8A.root")
mc_chain.Add("/eos/cms//store/mc/RunIISummer20UL18NanoAODv2/WJetsToLNu_TuneCP5_13TeV-madgraphMLM-pythia8/NANOAODSIM/106X_upgrade2018_realistic_v15_L1v1-v1/00000/8B47D4A5-572B-C24D-848A-A2945D0E794D.root")


## Data
data_chain = ROOT.TChain("Events")
print "--> Adding Data files to chain..."
# Dataset: /SingleMuon/Run2018D-UL2018_MiniAODv1_NanoAODv2-v1/NANOAOD
# List all files.
# dasgoclient -query="file dataset=/SingleMuon/Run2018D-UL2018_MiniAODv1_NanoAODv2-v1/NANOAOD"

data_chain.Add("/eos/cms/store/data/Run2018D/SingleMuon/NANOAOD/UL2018_MiniAODv1_NanoAODv2-v1/280000/0BF2854F-C799-2F41-A0F4-0274DC311372.root")


analyser.py
import tqdm 
import json
import ROOT
#.....
# setup some ROOT stuff
ROOT.PyConfig.DisableRootLogon = True
ROOT.PyConfig.IgnoreCommandLineOptions = False
ROOT.gROOT.SetBatch(True)

# Load source files chains
from source_files import data_chain 
from source_files import mc_chain 

# Config
max_events = 25000 # max number of events to be processed -> set to -1 to proccess all events
# max_events = -1 # max number of events to be processed -> set to -1 to proccess all events


# Declare histograms
# muons p_T
h_muon_pt_1 = ROOT.TH1F("h_muon_pt_1", "; Muon pT ; Muons", 50, 0, 200)
# force the storage and computation of the sum of the square of weights per bin
# REF: https://root.cern.ch/doc/master/classTH1.html#associated-errors
h_muon_pt_1.Sumw2()

h_muon_pt_2 = ROOT.TH1F("h_muon_pt_2", "; Muon pT ; Muons", 50, 0, 200)
# force the storage and computation of the sum of the square of weights per bin
# REF: https://root.cern.ch/doc/master/classTH1.html#associated-errors
h_muon_pt_2.Sumw2()

# muons phi
h_muon_phi_1 = ROOT.TH1F("h_muon_phi_1", "; Muon phi ; Muons", 50, -4, 4)
h_muon_phi_1.Sumw2()

h_muon_phi_2 = ROOT.TH1F("h_muon_phi_2", "; Muon phi ; Muons", 50, -4, 4)
h_muon_phi_2.Sumw2()

# MET Et
h_MET_phi_1 = ROOT.TH1F("h_MET_phi_1", "; MET phi ; Muons", 50, -5, 5)
h_MET_phi_1.Sumw2()

h_MET_phi_2 = ROOT.TH1F("h_MET_phi_2", "; MET phi ; Muons", 50, -5, 5)
h_MET_phi_2.Sumw2()

#MET pt
h_MET_pt_1 = ROOT.TH1F("h_MET_pt_1", "; MET pt ; Muons", 50, 0, 200)
h_MET_pt_1.Sumw2()

h_MET_pt_2 = ROOT.TH1F("h_MET_pt_2", "; MET pt ; Muons", 50, 0, 200)
h_MET_pt_2.Sumw2()

# number of muons per event
h_nmuon_1 = ROOT.TH1F("h_nmuon_1", "; Number of muons ; Events", 5, 0, 5)
h_nmuon_1.Sumw2() 

h_nmuon_2 = ROOT.TH1F("h_nmuon_2", "; Number of muons ; Events", 5, 0, 5)
h_nmuon_2.Sumw2() 

#invariant mass of muon and MET (CMS DATA)
h_MmM1 = ROOT.TH1F("h_MmM1","Transverse mass histogram (CMS DATA)", 50, 60, 200)
h_MmM1.Sumw2()

#invariant mass of muon and MET (Simulated DATA)
h_MmM2 = ROOT.TH1F("h_MmM2","Transverse mass histogram (MC DATA)", 50, 0, 200)
h_MmM2.Sumw2()

# invariant mass of two muons
h_Mmm = ROOT.TH1F("h_Mmm","Invariant mass of two muons", 50, 70, 100)
h_Mmm.Sumw2()



# total number of events in the file
n_events1 = data_chain.GetEntries()  #(REAL CMS DATA)

n_events2 = mc_chain.GetEntries()

# Set the number of events for percentage calc
n_events1_for_tqdm = n_events1
if max_events >= 0:
    n_events1_for_tqdm = max_events
    
n_events2_for_tqdm = int(n_events2*0.8/3.) # a correction factor applied in orther to not overflow the event list when running over all events
if max_events >= 0:
    n_events2_for_tqdm = max_events 


#calculating weight of MC data
# x-sec REF: 
# https://twiki.cern.ch/twiki/bin/viewauth/CMS/StandardModelCrossSectionsat13TeV
# https://pdglive.lbl.gov/Particle.action?node=S043&init=0

W_crosssection = 20508.9*1E-3 # in fb
br_W_to_leptons = 0.10785 #this is the mean value of Branching Ratio to muon+nu and electron+nu
luminosity = 1 # in fb^-1
W_mc = (W_crosssection*br_W_to_leptons*luminosity)/n_events2_for_tqdm

#weight for CMS data
W = 1

print "--> Weights:"
print "----> MC: " + str(W_mc) 
print "----> Data: "+ str(W)


# loop over events and fill histograms
counterA = 0
counterB = 0
counterC = 0
counterD = 0
counter1 = 0
count1 = 0

print "Processing CMS Data events..."

# Golden JSON File 2018
# /afs/cern.ch/cms/CAF/CMSCOMM/COMM_DQM/certification/Collisions18/13TeV/Legacy_2018/Cert_314472-325175_13TeV_Legacy2018_Collisions18_JSON.txt
def good_event(run,lumi):
    # Check whether this a good event
    with open('/afs/cern.ch/cms/CAF/CMSCOMM/COMM_DQM/certification/Collisions18/13TeV/Legacy_2018/Cert_314472-325175_13TeV_Legacy2018_Collisions18_JSON.txt') as json_file:
      JSONlist = json.load(json_file)

      if str(run) in JSONlist.keys():
          for rg in JSONlist[str(run)]:
              if len(rg) ==2:
                  if lumi>=rg[0] and lumi<=rg[1]:
                      return True
      
      return False

# Loop over Data events
for ievt, evt in tqdm.tqdm(enumerate(data_chain), desc ="Events"):     #(REAL CMS DATA)
    # only loop over good events - not even sure if it is needed...
    if good_event(evt.run, evt.luminosityBlock): 
        if ievt+1 > max_events and max_events > 0:
            break
        # loop over muons
        if evt.nMuon>=1:
          if evt.HLT_IsoMu24==1:
            for i_muon in range(evt.nMuon):
              if evt.Muon_pt > 26:
                if evt.MET_pt > 40: 
                  if evt.Muon_pfRelIso03_all[i_muon] < 0.35:
                    if abs(evt.Muon_eta[i_muon]) < 2.5:
                      muon1 = ROOT.Math.PtEtaPhiMVector(evt.Muon_pt[0], evt.Muon_eta[0], evt.Muon_phi[0], evt.Muon_mass[0])
                      MET1 = ROOT.Math.PtEtaPhiMVector(evt.MET_pt, 0, evt.MET_phi, 0)
                      TM_mM1 = muon1 + MET1
                      m_mM1 = TM_mM1.M()
                      if evt.Muon_tightId[i_muon]==1:
                        if (m_mM1 > 60 and m_mM1 < 200):
                          h_muon_pt_1.Fill(evt.Muon_pt[i_muon],W)
                          h_muon_phi_1.Fill(evt.Muon_phi[i_muon],W) 
                          h_MmM1.Fill(m_mM1,W)
                          h_MET_phi_1.Fill(evt.MET_phi,W)
                          h_MET_pt_1.Fill(evt.MET_pt,W)
                          h_nmuon_1.Fill(evt.nMuon,W)   # per event info
                          counterA+=W
                        else:
                          counterD+=W
                      else:
                        if (m_mM1 > 60 and m_mM1 < 200):
                          counterB+=W
                        else:
                          counterC+=W
                        counter1+=W
                      break
        
        count1+=W        
    
print("-->> Event counts - DATA")
print("Counter A: " + str(counterA))  
print("Counter B: " + str(counterB))
print("Counter C: " + str(counterC)) 
print("Counter D: " + str(counterD))
print("Counter 1 ???: " + str(counter1))
print("Total number of processed events: " + str(count1)) 


countera = 0
counterb = 0
counterc = 0
counterd = 0 
counter2 = 0
count2 = 0
     
print "Processing CMS MC events..."

# load PU weights file
pu_weight_file = ROOT.TFile('data/pu_weights_2018_from_HZZ4L.root')
histo_pu_weight = pu_weight_file.Get('weights')
histo_pu_weight_up = pu_weight_file.Get('weights_varUp')
histo_pu_weight_down = pu_weight_file.Get('weights_varDn')

def get_pu_weight(pu_true):
  pu_weight_nominal = histo_pu_weight.GetBinContent(histo_pu_weight.FindBin(pu_true))
  pu_weight_up = histo_pu_weight_up.GetBinContent(histo_pu_weight_up.FindBin(pu_true))
  pu_weight_down = histo_pu_weight_down.GetBinContent(histo_pu_weight_down.FindBin(pu_true))

  return pu_weight_nominal, pu_weight_up, pu_weight_down

# Loop over Data events
for ievt, evt in tqdm.tqdm(enumerate(mc_chain), desc ="Events"):
    # get PU weight and correct MC one
    pu_weight_nominal, pu_weight_up, pu_weight_down = get_pu_weight(evt.Pileup_nTrueInt)
    w_mc_effective = W_mc * pu_weight_nominal

    # filter only MC for muon + nu decay channel
    is_muon_decay = False
    for lhe_part in range(evt.nLHEPart):
      if abs(evt.LHEPart_pdgId[lhe_part]) == 13: # MUON DECAY
      # if abs(evt.LHEPart_pdgId[lhe_part]) == 11: # ELECTRON DECAY
        is_muon_decay = True

    if is_muon_decay:
      # condition to stop event processing, when max_events is reached
      if ievt+1 > max_events and max_events > 0:
          break

      # loop over muons 
      if evt.nMuon>=1:
        if evt.HLT_IsoMu24==1:         
          for i_muon in range(evt.nMuon):
            if evt.Muon_pt > 26:
              if evt.MET_pt > 40:
                if evt.Muon_pfRelIso03_all[i_muon] < 0.35:
                  if abs(evt.Muon_eta[i_muon]) < 2.5:     
                    muon2 = ROOT.Math.PtEtaPhiMVector(evt.Muon_pt[0], evt.Muon_eta[0], evt.Muon_phi[0], evt.Muon_mass[0])
                    MET2 = ROOT.Math.PtEtaPhiMVector(evt.MET_pt, 0, evt.MET_phi, 0)
                    TM_mM2 = muon2 + MET2
                    m_mM2 = TM_mM2.M()
                    if evt.Muon_tightId[i_muon]==1:
                      if (m_mM2 > 60 and m_mM2 < 200):
                        h_muon_pt_2.Fill(evt.Muon_pt[i_muon],w_mc_effective)
                        h_muon_phi_2.Fill(evt.Muon_phi[i_muon],w_mc_effective) 
                        h_MmM2.Fill(m_mM2,w_mc_effective)
                        h_MET_phi_2.Fill(evt.MET_phi,w_mc_effective)
                        h_MET_pt_2.Fill(evt.MET_pt,w_mc_effective)
                        h_nmuon_2.Fill(evt.nMuon,w_mc_effective)
                        countera+=w_mc_effective
                      else:
                        counterd+=w_mc_effective
                    else:
                      if (m_mM2 > 60 and m_mM2 < 200):
                        counterb+=w_mc_effective
                      else:
                        counterc+=w_mc_effective
                      counter2+=w_mc_effective
                    break
                  
      count2+=w_mc_effective                   

print("-->> Event counts (weighted) - MC")
print("Counter A: " + str(countera))  
print("Counter B: " + str(counterb))
print("Counter C: " + str(counterc)) 
print("Counter D: " + str(counterd))
print("Counter 2 ???: " + str(counter2))
print("Total number of processed events: " + str(count2)) 

    
          
# Save histogram to a file
# Open a ROOT file and save the formula, function and histogram
output_file = ROOT.TFile( 'output_file.root', 'RECREATE')

# Write histograms to file
h_muon_pt_1.Write()
h_muon_phi_1.Write()
h_MET_phi_1.Write()
h_MET_pt_1.Write()
h_nmuon_1.Write()
h_muon_pt_2.Write()
h_muon_phi_2.Write()
h_MET_phi_2.Write()
h_MET_pt_2.Write()
h_nmuon_2.Write()
h_MmM1.Write()
h_MmM2.Write()


# Close file
output_file.Close()

print "\nOutput histograms have been saved output_file.root"
